### PR TITLE
fix: staking cache to in-memory KV store

### PR DESCRIPTION
**Pull request by randy-cro**
## Description

This PR fixes inconsistencies in the staking cache (originally implemented in [PR #1725](https://github.com/crypto-org-chain/cosmos-sdk/pull/1725)) across different nodes.
Previously, simulation or optimistic execution contexts did not properly handle discarded changes, which could lead to duplicate or incorrect cache data, eventually causing apphash mismatches and halting the chain.

The primary change refactors the cache to use an **in-memory KV store.**
With this approach, when a context is discarded, the cache is automatically rolled back, just like any other store in the system.

This ensures that cache state remains consistent with the underlying store, even during simulations or optimistic execution scenarios.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] added `!` to the type prefix if API or client breaking change
* [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
* [ ] provided a link to the relevant issue or specification
* [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules)
* [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
* [ ] added a changelog entry to `CHANGELOG.md`
* [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
* [ ] updated the relevant documentation or specification
* [ ] reviewed "Files changed" and left comments if necessary
* [ ] run `make lint` and `make test`
* [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed `!` in the type prefix if API or client breaking change
* [ ] confirmed all author checklist items have been addressed 
* [ ] reviewed state machine logic
* [ ] reviewed API design and naming
* [ ] reviewed documentation is accurate
* [ ] reviewed tests and test coverage
* [ ] manually tested (if applicable)

### CODE REVIEW

File: baseapp/abci.go [hunk 1]
```diff
@@ -911,6 +911,7 @@ func (app *BaseApp) FinalizeBlock(req *abci.RequestFinalizeBlock) (res *abci.Res
 	}()
 
 	if app.optimisticExec.Initialized() {
+		app.logger.Error("BASEAPP::::FinalizeBlock: Using optimistic execution", "height", req.Height)
```
Comment (JayT106) [inline baseapp/abci.go]: should be Debug level log, you can also remove `BASEAPP::::` , the log should add the baseapp module name to the log automatically.

File: baseapp/abci.go [hunk 2]
```diff
@@ -919,21 +920,27 @@ func (app *BaseApp) FinalizeBlock(req *abci.RequestFinalizeBlock) (res *abci.Res
 		// only return if we are not aborting
 		if !aborted {
 			if res != nil {
+				app.logger.Error("BASEAPP::::FinalizeBlock (OE): About to compute app hash", "height", req.Height)
```
Comment (JayT106) [inline baseapp/abci.go]: can remove this log, already print the info later

**Review by JayT106 — COMMENTED** [review]
LGTM, but the log info and level should be refined and simplified, more of it should put it into debug level. Some logs are using error level which the code section doesn't handle the error case.

File: store/rootmulti/store.go [hunk 3]
```diff
@@ -564,7 +564,23 @@ func (rs *Store) WorkingHash() []byte {
 		return storeInfos[i].Name < storeInfos[j].Name
 	})
 
-	return types.CommitInfo{StoreInfos: storeInfos}.Hash()
+	appHash := types.CommitInfo{StoreInfos: storeInfos}.Hash()
+
+	// Log all individual store hashes for debugging
```
Comment (songgaoye) [inline store/rootmulti/store.go]: as @JayT106 said it's better to use logger.Debug

**Comment by songgaoye** [comment]
https://github.com/crypto-org-chain/cosmos-sdk/blob/6fa6924b2d346db588c4cade4f73c79e998ade80/server/start.go#L1003
I saw that the `FlagStakingCacheSize` default value is 0 (unlimited) and also didn't see in config.toml. Do you want to use unlimited cache capacity? @randy-cro

File: x/staking/cache/cache.go [hunk 4]
```diff
@@ -2,170 +2,355 @@ package cache
...
+func (e *Entry[V, E]) setEntry(ctx context.Context, cdc codec.BinaryCodec, key string, value V) error {
 	e.mu.Lock()
 	defer e.mu.Unlock()
+	return e.setEntryUnsafe(ctx, cdc, key, value)
+}
 
+// setEntryUnsafe works the same as setEntry but the caller is responsible for holding the lock
+func (e *Entry[V, E]) setEntryUnsafe(ctx context.Context, cdc codec.BinaryCodec, key string, value V) error {
 	if e.full.Load() {
-		return
```
Comment (songgaoye) [inline x/staking/cache/cache.go]: there’s an edge case: if the entry is marked full but the key already exists, should its value still be updated? I think we also need to add unit test to describe this behavior.

Comment (randy-cro) [reply x/staking/cache/cache.go]: in my opinion, this is indeed an edge case but ive decided not to allow update to simplify the code. This is because the setEntry happens based on blockTime, so there usually wouldnt be an update unless there are 2 unbonding transactions in one block. Even if that happens, one entry will be in the cache will the other will not. Regardless, at this point the cache will be considered stale and because the deletion of the entry is based on key(timestamp-dependent), the key will still be deleted once the 2 unbonding delegations are mature.

**Comment by randy-cro** [comment]
> Member

image
image


@songgaoye it is here, did u miss it?

File: CHANGELOG.md [hunk 5]
```diff
@@ -40,6 +40,9 @@ Ref: https://keepachangelog.com/en/1.0.0/
 
 ### Features
 
+* (x/staking) [#1733](https://github.com/crypto-org-chain/cosmos-sdk/pull/1730) Fix staking cache to use an in-memory KV store to ensure rollback on discarded contexts.
```
Comment (JayT106) [inline CHANGELOG.md]: The link is duplicate to the next one.

**Review by JayT106 — APPROVED** [review]
LGTM

File: CHANGELOG.md [hunk 6]
```diff
@@ -40,6 +40,9 @@ Ref: https://keepachangelog.com/en/1.0.0/
...
+* (x/staking) [#1730](https://github.com/crypto-org-chain/cosmos-sdk/pull/1730) Normalize cache validator queue key to be UTC.
+* (x/staking) [#1725](https://github.com/crypto-org-chain/cosmos-sdk/pull/1725) Optimize staking endblocker execution by caching queue entries from iterators.
```
Comment (thomas-nguy) [inline CHANGELOG.md]: Lets just group to a single changelog with a link to this PR for better clarify

File: x/staking/keeper/validator.go [hunk 7]
```diff
@@ -474,7 +482,18 @@ func (k Keeper) SetUnbondingValidatorsQueue(ctx context.Context, endTime time.Ti
...
-	return store.Set(types.GetValidatorQueueKey(endTime, endHeight), bz)
+	err = store.Set(types.GetValidatorQueueKey(endTime, endHeight), bz)
+	if err != nil {
+		return err
+	}
+
+	if k.cache != nil {
+		err = k.cache.SetUnbondingValidatorsQueue(ctx, types.GetCacheValidatorQueueKey(endTime, endHeight), addrs)
+		if err != nil {
+			k.Logger(ctx).Error("SetUnbondingValidatorsQueue to cache failed", "error", err)
```
Comment (thomas-nguy) [inline x/staking/keeper/validator.go:496]: nitpicking but in case of Error, it would be better to disable the cache, such as setting it to Nil to avoid strange behavior

Unless its ErrCacheMaxSizeReached

File: x/staking/cache/entry.go [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+
+	var metadata types.CacheMetadata
+	if err := cdc.Unmarshal(bz, &metadata); err != nil {
+		return types.CacheMetadata{}, err
+	}
+	return metadata, nil
+}
+
+// setMetadata: caller MUST hold lock
+func (e *Entry[V, E]) setMetadata(store corestoretypes.KVStore, cdc codec.BinaryCodec, metadata types.CacheMetadata) error {
```
Comment (thomas-nguy) [inline x/staking/cache/entry.go]: nitpicky: the metadata structure may be an overkill because you are introducing unnecessary  marshalling and unmarshaling everytime

File: x/staking/cache/entry.go [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+			return err
+		}
+	}
+
+	bz, err := marshal(cdc, e.entryType, value)
+	if err != nil {
+		return err
+	}
+
+	if err := e.setStore(store, key, bz); err != nil {
```
Comment (thomas-nguy) [inline x/staking/cache/entry.go]: Dont you have to check if its full yet before set in store?

Comment (randy-cro) [reply x/staking/cache/entry.go]: good catch, fixed

File: x/staking/cache/entry.go (hunk 8 above)
Comment (randy-cro) [reply x/staking/cache/entry.go]: the metadata structure i created because if i were to use normal booleans it will not be discarded in the case of a OE abort/simulation

File: x/staking/cache/entry.go [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	store := e.storeService.OpenMemoryStore(ctx)
+	exist := false
+	if e.max > 0 {
+		var err error
+		exist, err = e.exists(store, key)
+		if err != nil {
+			return err
```
Comment (thomas-nguy) [inline x/staking/cache/entry.go:193]: In case the cache is full, the entry may not be there, and if you return an error, it will permanently disable the cache

File: x/staking/cache/entry.go [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+// setUnsafe: caller MUST hold lock
+func (e *Entry[V, E]) setUnsafe(ctx context.Context, cdc codec.BinaryCodec, key string, value V) error {
+	store := e.storeService.OpenMemoryStore(ctx)
+
+	metadata, err := e.getMetadata(store, cdc)
+	if err != nil {
+		return err
+	}
+
+	if metadata.IsFull {
```
Comment (thomas-nguy) [inline x/staking/cache/entry.go:330]: can use `e. isFull`

File: x/staking/cache/entry.go [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+		return types.ErrCacheMaxSizeReached
+	}
+
+	exist := false
+	if e.max > 0 {
+		var err error
+		exist, err = e.exists(store, key)
```
Comment (thomas-nguy) [inline x/staking/cache/entry.go]: Not sure what are you checking there. can't you set the same key twice?

File: x/staking/cache/entry.go (hunk 11 above)
Comment (randy-cro) [reply x/staking/cache/entry.go:330]: e.isFull will need to hold the rlock, will cause a dead lock as the caller of setUnsafe will be holding the lock already

File: x/staking/cache/entry.go (hunk 10 above)
Comment (randy-cro) [reply x/staking/cache/entry.go:193]: image


it is to set the cache as full = false once the entry is deleted

Comment (randy-cro) [reply x/staking/cache/entry.go:193]: if the entry is not there `exist` will just be false?

File: x/staking/cache/entry.go (hunk 12 above)
Comment (randy-cro) [reply x/staking/cache/entry.go]: image

again is to optimize the counting, if the key already exist and its not full,`set` will just be a replace so no need to count again

File: x/staking/cache/cache.go [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+	}
+
+	c.logger(ctx).Info("Unbonding validators queue is dirty. Reinitializing cache from store.")
+
+	data, err := c.unbondingValidatorsQueue.loadFromStore(ctx)
+	if err != nil {
+		return err
+	}
+
+	if err := c.unbondingValidatorsQueue.clear(ctx, c.cdc); err != nil {
```
Comment (thomas-nguy) [inline x/staking/cache/cache.go]: Wonder if it is faster just to create a newEntry and let the garbage collector clean it for you

File: x/staking/cache/cache.go [hunk 14]
```diff
@@ -0,0 +1,299 @@
...
+		return err
+	}
+
+	for key, value := range data {
+		if err := c.unbondingValidatorsQueue.setUnsafe(ctx, c.cdc, key, value); err != nil {
```
Comment (thomas-nguy) [inline x/staking/cache/cache.go]: Maybe that one can be optimized to open store service one and skip the meta data check

File: x/staking/cache/cache.go (hunk 13 above)
Comment (thomas-nguy) [reply x/staking/cache/cache.go]: same for the other function `checkReloadUnbondingDelegationsQueue` etc

File: x/staking/cache/cache.go (hunk 14 above)
Comment (thomas-nguy) [reply x/staking/cache/cache.go]: same for the other functions `checkReloadUnbondingDelegationsQueue` etc

File: x/staking/cache/cache.go [hunk 15]
```diff
@@ -0,0 +1,299 @@
...
+
+	if err := c.checkReloadUnbondingValidatorsQueue(ctx); err != nil {
+		return nil, err
+	}
+
+	return c.unbondingValidatorsQueue.get(ctx, c.cdc, types.GetCacheValidatorQueueKey(endTime, endHeight))
+}
+
+func (c *ValidatorsQueueCache) SetUnbondingValidatorsQueue(ctx context.Context, key string, addrs []string) error {
+	return c.unbondingValidatorsQueue.set(ctx, c.cdc, key, addrs)
```
Comment (thomas-nguy) [inline x/staking/cache/cache.go:80]: nitpicky:  In case of Get, you are managing the cache size `ErrCacheMaxSizeReached` at the cache level, but in case of Set and Delete , you are managing it at the Entry level.  Its not very consistent

File: x/staking/cache/cache.go (hunk 14 above)
Comment (randy-cro) [reply x/staking/cache/cache.go]: what do u mean by this?

File: x/staking/cache/cache.go (hunk 15 above)
Comment (randy-cro) [reply x/staking/cache/cache.go:80]: refactored

File: x/staking/cache/entry.go (hunk 10 above)
Comment (thomas-nguy) [reply x/staking/cache/entry.go:193]: ok nvm

File: x/staking/cache/cache.go (hunk 13 above)
Comment (randy-cro) [reply x/staking/cache/cache.go]: hmm think the implementation will become more complex because all 3 entry now are sharing the same memStoreService, one entry full doesnt mean another is full

File: x/staking/cache/entry.go [hunk 16]
```diff
@@ -0,0 +1,344 @@
...
+// checkReload: caller MUST hold lock
+func (e *Entry[V, E]) checkReload(ctx context.Context, cdc codec.BinaryCodec, logger func(ctx context.Context) log.Logger) error {
+	store := e.storeService.OpenMemoryStore(ctx)
+
+	metadata, err := e.getMetadata(store, cdc)
+	if err != nil {
+		return err
+	}
+
+	if metadata.IsFull {
```
Comment (thomas-nguy) [inline x/staking/cache/entry.go]: Can you explain this part? Shouldnt it be metadata.IsDirty? And Why this error?

File: x/staking/cache/entry.go [hunk 17]
```diff
@@ -0,0 +1,344 @@
...
+	}
+
+	if e.max > 0 && !exist {
+		count, err := e.count(store)
+		if err != nil {
+			return err
+		}
+
+		if count >= uint64(e.max) {
+			metadata.IsDirty = true
```
Comment (thomas-nguy) [inline x/staking/cache/entry.go]: Technically The cache is not `Dirty` when it reach `Max`, but it become dirty the next time the `Set` is called and when it is already Full, in that case you should skip the set and just set Dirty to True

Otherwise those two booleans do the same thing

File: x/staking/cache/entry.go [hunk 18]
```diff
@@ -0,0 +1,344 @@
...
+	}
+
+	data, err := e.loadFromStore(ctx)
+	if err != nil {
+		return err
+	}
+
+	size := len(data)
+	if e.max > 0 && size >= int(e.max) {
+		metadata.IsFull = true
```
Comment (thomas-nguy) [inline x/staking/cache/entry.go]: Maybe change to IsDirty = true if it exceed the max size and return

and IsFull = true if it exactly reach max size

File: x/staking/cache/entry.go (hunk 16 above)
Comment (randy-cro) [reply x/staking/cache/entry.go]: this part is basically to short circuit and return max error when the cache is already full, probably should be taken out and be placed in `get` and `getAll` to be clearer?

File: x/staking/cache/cache.go (hunk 4 above)
Comment (randy-cro) [reply x/staking/cache/cache.go]: @songgaoye added this functionality as suggested by you, also added unit test to test that it works!

**Review by thomas-nguy — APPROVED** [review]
